File: minivault.py
#!/usr/bin/env python3
"""
MiniVault - Encrypted .env secret manager
Version: 1.0.0
"""

import sys
import os
import hashlib
import base64
import json
import signal
import getpass
import glob
import struct
from pathlib import Path

try:
    from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
    from cryptography.hazmat.backends import default_backend
    USE_CRYPTOGRAPHY = True
except ImportError:
    USE_CRYPTOGRAPHY = False

# Fallback: use subprocess with openssl binary
import subprocess

# ──────────────────────────────────────────────
# Crypto Standard (matches PHP & Node versions)
#   Algorithm : AES-256-CBC
#   Key derivation : SHA-256  (32 bytes)
#   IV : 16 random bytes
#   Integrity : SHA-256 of plaintext (32 bytes)
#   Wire format : base64( IV[16] + INTEGRITY[32] + CIPHERTEXT )
# ──────────────────────────────────────────────

MINIVAULT_EXT = ".enc"
STATE_EXT = ".mvstate"


def derive_key(password: str) -> bytes:
    return hashlib.sha256(password.encode("utf-8")).digest()


def encrypt(plaintext: str, password: str) -> str:
    key = derive_key(password)
    iv = os.urandom(16)
    integrity = hashlib.sha256(plaintext.encode("utf-8")).digest()  # 32 bytes
    data = plaintext.encode("utf-8")

    # PKCS7 padding
    pad_len = 16 - (len(data) % 16)
    data += bytes([pad_len] * pad_len)

    if USE_CRYPTOGRAPHY:
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        encryptor = cipher.encryptor()
        ciphertext = encryptor.update(data) + encryptor.finalize()
    else:
        ciphertext = _openssl_encrypt(key, iv, data)

    wire = base64.b64encode(iv + integrity + ciphertext).decode("ascii")
    return wire


def decrypt(wire: str, password: str) -> str:
    raw = base64.b64decode(wire)
    if len(raw) < 48:
        raise ValueError("Archivo cifrado corrupto o demasiado corto.")
    iv = raw[:16]
    integrity = raw[16:48]
    ciphertext = raw[48:]
    key = derive_key(password)

    if USE_CRYPTOGRAPHY:
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        decryptor = cipher.decryptor()
        padded = decryptor.update(ciphertext) + decryptor.finalize()
    else:
        padded = _openssl_decrypt(key, iv, ciphertext)

    # Remove PKCS7 padding
    pad_len = padded[-1]
    if pad_len < 1 or pad_len > 16:
        raise ValueError("Contraseña incorrecta o archivo corrupto.")
    plaintext_bytes = padded[:-pad_len]

    # Verify integrity
    computed = hashlib.sha256(plaintext_bytes).digest()
    if computed != integrity:
        raise ValueError("Fallo de integridad: la contraseña es incorrecta o el archivo fue modificado.")

    return plaintext_bytes.decode("utf-8")


# ── OpenSSL subprocess fallback ────────────────

def _openssl_encrypt(key: bytes, iv: bytes, padded_data: bytes) -> bytes:
    proc = subprocess.run(
        ["openssl", "enc", "-aes-256-cbc", "-nosalt", "-nopad",
         "-K", key.hex(), "-iv", iv.hex()],
        input=padded_data, capture_output=True
    )
    if proc.returncode != 0:
        raise RuntimeError("openssl enc falló: " + proc.stderr.decode())
    return proc.stdout


def _openssl_decrypt(key: bytes, iv: bytes, ciphertext: bytes) -> bytes:
    proc = subprocess.run(
        ["openssl", "enc", "-d", "-aes-256-cbc", "-nosalt", "-nopad",
         "-K", key.hex(), "-iv", iv.hex()],
        input=ciphertext, capture_output=True
    )
    if proc.returncode != 0:
        raise RuntimeError("openssl dec falló: " + proc.stderr.decode())
    return proc.stdout


# ── State file ─────────────────────────────────

def _state_path(enc_path: str) -> str:
    return enc_path + STATE_EXT


def _save_state(enc_path: str, tmp_path: str, content_hash: str):
    state = {"enc": enc_path, "tmp": tmp_path, "hash": content_hash}
    with open(_state_path(enc_path), "w") as f:
        json.dump(state, f)


def _load_state(enc_path: str):
    sp = _state_path(enc_path)
    if not os.path.exists(sp):
        return None
    with open(sp) as f:
        return json.load(f)


def _delete_state(enc_path: str):
    sp = _state_path(enc_path)
    if os.path.exists(sp):
        os.remove(sp)


def _file_hash(path: str) -> str:
    with open(path, "rb") as f:
        return hashlib.sha256(f.read()).hexdigest()


def _secure_delete(path: str):
    """Overwrite with zeros before unlinking."""
    if not os.path.exists(path):
        return
    try:
        size = os.path.getsize(path)
        with open(path, "r+b") as f:
            f.write(b"\x00" * size)
            f.flush()
            os.fsync(f.fileno())
    except Exception:
        pass
    os.remove(path)


# ── File selection helpers ──────────────────────

def _find_env_files(directory: str = ".") -> list:
    return glob.glob(os.path.join(directory, "*.env")) + glob.glob(os.path.join(directory, ".env"))


def _find_enc_files(directory: str = ".") -> list:
    return glob.glob(os.path.join(directory, f"*{MINIVAULT_EXT}"))


def _select_file(files: list, label: str) -> str:
    if len(files) == 1:
        print(f"  Usando: {files[0]}")
        return files[0]
    print(f"\n  Se encontraron múltiples archivos {label}:")
    for i, f in enumerate(files):
        print(f"  [{i + 1}] {f}")
    while True:
        choice = input("  Seleccioná un número: ").strip()
        if choice.isdigit() and 1 <= int(choice) <= len(files):
            return files[int(choice) - 1]
        print("  Opción inválida.")


def _ask_password(confirm: bool = False) -> str:
    while True:
        pwd = getpass.getpass("  Contraseña: ")
        if not pwd:
            print("  La contraseña no puede estar vacía.")
            continue
        if confirm:
            pwd2 = getpass.getpass("  Confirmar contraseña: ")
            if pwd != pwd2:
                print("  Las contraseñas no coinciden.")
                continue
        return pwd


# ── Commands ───────────────────────────────────

def cmd_create(args: list):
    """Cifra un archivo .env existente."""
    print("\n🔒 MiniVault — Modo CREACIÓN\n")

    # Determine source .env
    if args:
        src = args[0]
        if not os.path.exists(src):
            print(f"  Error: no existe '{src}'")
            sys.exit(1)
    else:
        files = _find_env_files()
        if not files:
            print("  No se encontró ningún archivo .env en el directorio actual.")
            sys.exit(1)
        src = _select_file(files, ".env")

    # Determine destination
    stem = Path(src).stem if not src.endswith(".env") else Path(src).name
    dest = stem.lstrip(".") + MINIVAULT_EXT
    if args and len(args) > 1:
        dest = args[1]

    password = _ask_password(confirm=True)

    with open(src, "r", encoding="utf-8") as f:
        plaintext = f.read()

    wire = encrypt(plaintext, password)
    with open(dest, "w") as f:
        f.write(wire)

    print(f"\n  ✅ Archivo cifrado guardado en: {dest}")


def cmd_open(args: list):
    """Descifra temporalmente un archivo .enc para edición."""
    print("\n🔓 MiniVault — Modo APERTURA\n")

    # Determine source .enc
    if args:
        enc_path = args[0]
        if not os.path.exists(enc_path):
            print(f"  Error: no existe '{enc_path}'")
            sys.exit(1)
    else:
        files = _find_enc_files()
        if not files:
            print(f"  No se encontró ningún archivo {MINIVAULT_EXT} en el directorio actual.")
            sys.exit(1)
        enc_path = _select_file(files, MINIVAULT_EXT)

    # Determine temp .env path
    stem = Path(enc_path).stem
    tmp_path = stem + ".env"

    password = _ask_password(confirm=False)

    # Read and decrypt
    with open(enc_path, "r") as f:
        wire = f.read().strip()

    try:
        plaintext = decrypt(wire, password)
    except ValueError as e:
        print(f"\n  ❌ {e}")
        sys.exit(1)

    # Write temp file
    with open(tmp_path, "w", encoding="utf-8") as f:
        f.write(plaintext)

    initial_hash = _file_hash(tmp_path)
    _save_state(enc_path, tmp_path, initial_hash)

    print(f"\n  ✅ Archivo temporal: {tmp_path}")
    print("  Editá el archivo. Cuando termines, presioná Enter para cerrar la sesión.")
    print("  (Ctrl+C también cierra de forma segura)\n")

    # Register cleanup on SIGINT
    def cleanup(sig=None, frame=None):
        _close_session(enc_path, tmp_path, initial_hash, password)
        sys.exit(0)

    signal.signal(signal.SIGINT, cleanup)
    signal.signal(signal.SIGTERM, cleanup)

    input()
    _close_session(enc_path, tmp_path, initial_hash, password)


def _close_session(enc_path: str, tmp_path: str, initial_hash: str, password: str):
    print("\n  Cerrando sesión...")

    if not os.path.exists(tmp_path):
        print("  El archivo temporal ya no existe.")
        _delete_state(enc_path)
        return

    current_hash = _file_hash(tmp_path)

    if current_hash != initial_hash:
        print("  Se detectaron cambios. Actualizando archivo cifrado...")
        with open(tmp_path, "r", encoding="utf-8") as f:
            new_content = f.read()
        wire = encrypt(new_content, password)
        with open(enc_path, "w") as f:
            f.write(wire)
        print(f"  ✅ {enc_path} actualizado.")
    else:
        print("  Sin cambios. El archivo cifrado no fue modificado.")

    _secure_delete(tmp_path)
    _delete_state(enc_path)
    print(f"  🗑️  Archivo temporal eliminado. Exposición: 0.")


# ── Entry point ────────────────────────────────

USAGE = """
MiniVault v1.0.0 — Gestor seguro de secretos .env

Uso:
  python minivault.py create [archivo.env] [destino.enc]
  python minivault.py open   [archivo.enc]

Comandos:
  create    Cifra un archivo .env → genera un .enc
  open      Descifra temporalmente un .enc para edición

Sin argumentos de archivo, escanea el directorio actual automáticamente.
"""


def main():
    if len(sys.argv) < 2 or sys.argv[1] in ("-h", "--help"):
        print(USAGE)
        sys.exit(0)

    command = sys.argv[1].lower()
    rest = sys.argv[2:]

    if command == "create":
        cmd_create(rest)
    elif command == "open":
        cmd_open(rest)
    else:
        print(f"  Comando desconocido: '{command}'")
        print(USAGE)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: minivault.php
#!/usr/bin/env php
<?php
/**
 * MiniVault - Encrypted .env secret manager
 * Version: 1.0.0
 *
 * Crypto Standard:
 *   Algorithm  : AES-256-CBC
 *   Key        : SHA-256(password) → 32 bytes
 *   IV         : 16 random bytes
 *   Integrity  : SHA-256(plaintext) → 32 bytes
 *   Wire format: base64( IV[16] + INTEGRITY[32] + CIPHERTEXT )
 */

declare(strict_types=1);

define('MINIVAULT_EXT', '.enc');
define('STATE_EXT',     '.mvstate');

// ── Crypto ────────────────────────────────────

function derive_key(string $password): string
{
    return hash('sha256', $password, true); // raw binary, 32 bytes
}

function mv_encrypt(string $plaintext, string $password): string
{
    $key       = derive_key($password);
    $iv        = random_bytes(16);
    $integrity = hash('sha256', $plaintext, true); // 32 bytes

    $ciphertext = openssl_encrypt(
        $plaintext,
        'aes-256-cbc',
        $key,
        OPENSSL_RAW_DATA,
        $iv
    );

    if ($ciphertext === false) {
        throw new \RuntimeException('openssl_encrypt falló: ' . openssl_error_string());
    }

    return base64_encode($iv . $integrity . $ciphertext);
}

function mv_decrypt(string $wire, string $password): string
{
    $raw = base64_decode($wire, true);
    if ($raw === false || strlen($raw) < 48) {
        throw new \RuntimeException('Archivo cifrado corrupto o demasiado corto.');
    }

    $iv         = substr($raw, 0, 16);
    $integrity  = substr($raw, 16, 32);
    $ciphertext = substr($raw, 48);
    $key        = derive_key($password);

    $plaintext = openssl_decrypt(
        $ciphertext,
        'aes-256-cbc',
        $key,
        OPENSSL_RAW_DATA,
        $iv
    );

    if ($plaintext === false) {
        throw new \RuntimeException('Contraseña incorrecta o archivo corrupto.');
    }

    $computed = hash('sha256', $plaintext, true);
    if (!hash_equals($computed, $integrity)) {
        throw new \RuntimeException('Fallo de integridad: la contraseña es incorrecta o el archivo fue modificado.');
    }

    return $plaintext;
}

// ── State ─────────────────────────────────────

function state_path(string $enc_path): string
{
    return $enc_path . STATE_EXT;
}

function save_state(string $enc_path, string $tmp_path, string $hash): void
{
    file_put_contents(
        state_path($enc_path),
        json_encode(['enc' => $enc_path, 'tmp' => $tmp_path, 'hash' => $hash])
    );
}

function load_state(string $enc_path): ?array
{
    $sp = state_path($enc_path);
    if (!file_exists($sp)) return null;
    return json_decode(file_get_contents($sp), true);
}

function delete_state(string $enc_path): void
{
    $sp = state_path($enc_path);
    if (file_exists($sp)) unlink($sp);
}

function file_hash_mv(string $path): string
{
    return hash_file('sha256', $path);
}

function secure_delete(string $path): void
{
    if (!file_exists($path)) return;
    $size = filesize($path);
    $fh   = fopen($path, 'r+b');
    if ($fh) {
        fwrite($fh, str_repeat("\x00", $size));
        fflush($fh);
        fclose($fh);
    }
    unlink($path);
}

// ── Helpers ───────────────────────────────────

function find_env_files(string $dir = '.'): array
{
    $found = [];
    foreach (scandir($dir) as $f) {
        if ($f === '.env' || (str_ends_with($f, '.env') && $f !== '.' && $f !== '..')) {
            $found[] = $dir . DIRECTORY_SEPARATOR . $f;
        }
    }
    return $found;
}

function find_enc_files(string $dir = '.'): array
{
    $found = [];
    foreach (scandir($dir) as $f) {
        if (str_ends_with($f, MINIVAULT_EXT)) {
            $found[] = $dir . DIRECTORY_SEPARATOR . $f;
        }
    }
    return $found;
}

function select_file(array $files, string $label): string
{
    if (count($files) === 1) {
        echo "  Usando: {$files[0]}\n";
        return $files[0];
    }
    echo "\n  Se encontraron múltiples archivos {$label}:\n";
    foreach ($files as $i => $f) {
        echo "  [" . ($i + 1) . "] {$f}\n";
    }
    while (true) {
        $choice = trim(fgets(STDIN));
        if (ctype_digit($choice) && (int)$choice >= 1 && (int)$choice <= count($files)) {
            return $files[(int)$choice - 1];
        }
        echo "  Opción inválida.\n";
    }
}

function ask_password(bool $confirm = false): string
{
    while (true) {
        // Try to hide input on Unix
        if (DIRECTORY_SEPARATOR === '/') {
            system('stty -echo');
        }
        echo '  Contraseña: ';
        $pwd = trim(fgets(STDIN));
        if (DIRECTORY_SEPARATOR === '/') {
            system('stty echo');
            echo "\n";
        }
        if ($pwd === '') {
            echo "  La contraseña no puede estar vacía.\n";
            continue;
        }
        if (!$confirm) return $pwd;

        if (DIRECTORY_SEPARATOR === '/') system('stty -echo');
        echo '  Confirmar contraseña: ';
        $pwd2 = trim(fgets(STDIN));
        if (DIRECTORY_SEPARATOR === '/') {
            system('stty echo');
            echo "\n";
        }
        if ($pwd !== $pwd2) {
            echo "  Las contraseñas no coinciden.\n";
            continue;
        }
        return $pwd;
    }
}

// ── Commands ──────────────────────────────────

function cmd_create(array $args): void
{
    echo "\n🔒 MiniVault — Modo CREACIÓN\n\n";

    if (!empty($args[0])) {
        $src = $args[0];
        if (!file_exists($src)) {
            echo "  Error: no existe '{$src}'\n";
            exit(1);
        }
    } else {
        $files = find_env_files();
        if (empty($files)) {
            echo "  No se encontró ningún archivo .env en el directorio actual.\n";
            exit(1);
        }
        $src = select_file($files, '.env');
    }

    $stem = preg_replace('/^\./', '', pathinfo($src, PATHINFO_FILENAME));
    $dest = $args[1] ?? ($stem . MINIVAULT_EXT);

    $password  = ask_password(confirm: true);
    $plaintext = file_get_contents($src);
    $wire      = mv_encrypt($plaintext, $password);
    file_put_contents($dest, $wire);
    echo "\n  ✅ Archivo cifrado guardado en: {$dest}\n";
}

function cmd_open(array $args): void
{
    echo "\n🔓 MiniVault — Modo APERTURA\n\n";

    if (!empty($args[0])) {
        $enc_path = $args[0];
        if (!file_exists($enc_path)) {
            echo "  Error: no existe '{$enc_path}'\n";
            exit(1);
        }
    } else {
        $files = find_enc_files();
        if (empty($files)) {
            echo "  No se encontró ningún archivo " . MINIVAULT_EXT . " en el directorio actual.\n";
            exit(1);
        }
        $enc_path = select_file($files, MINIVAULT_EXT);
    }

    $stem     = basename($enc_path, MINIVAULT_EXT);
    $tmp_path = dirname($enc_path) . DIRECTORY_SEPARATOR . $stem . '.env';

    $password = ask_password(confirm: false);
    $wire     = trim(file_get_contents($enc_path));

    try {
        $plaintext = mv_decrypt($wire, $password);
    } catch (\RuntimeException $e) {
        echo "\n  ❌ " . $e->getMessage() . "\n";
        exit(1);
    }

    file_put_contents($tmp_path, $plaintext);
    $initial_hash = file_hash_mv($tmp_path);
    save_state($enc_path, $tmp_path, $initial_hash);

    echo "\n  ✅ Archivo temporal: {$tmp_path}\n";
    echo "  Editá el archivo. Cuando termines, presioná Enter para cerrar la sesión.\n";
    echo "  (Ctrl+C también cierra de forma segura)\n\n";

    // Cleanup function
    $close_session = function () use ($enc_path, $tmp_path, $initial_hash, $password): void {
        echo "\n  Cerrando sesión...\n";
        if (!file_exists($tmp_path)) {
            echo "  El archivo temporal ya no existe.\n";
            delete_state($enc_path);
            return;
        }
        $current_hash = file_hash_mv($tmp_path);
        if ($current_hash !== $initial_hash) {
            echo "  Se detectaron cambios. Actualizando archivo cifrado...\n";
            $new_content = file_get_contents($tmp_path);
            $new_wire    = mv_encrypt($new_content, $password);
            file_put_contents($enc_path, $new_wire);
            echo "  ✅ {$enc_path} actualizado.\n";
        } else {
            echo "  Sin cambios. El archivo cifrado no fue modificado.\n";
        }
        secure_delete($tmp_path);
        delete_state($enc_path);
        echo "  🗑️  Archivo temporal eliminado. Exposición: 0.\n";
    };

    // Register signal handlers
    if (function_exists('pcntl_signal')) {
        pcntl_signal(SIGINT,  function () use ($close_session) { $close_session(); exit(0); });
        pcntl_signal(SIGTERM, function () use ($close_session) { $close_session(); exit(0); });
        pcntl_async_signals(true);
    }

    fgets(STDIN); // Wait for Enter
    $close_session();
}

// ── Main ──────────────────────────────────────

$usage = <<<EOT

MiniVault v1.0.0 — Gestor seguro de secretos .env

Uso:
  php minivault.php create [archivo.env] [destino.enc]
  php minivault.php open   [archivo.enc]

Comandos:
  create    Cifra un archivo .env → genera un .enc
  open      Descifra temporalmente un .enc para edición

Sin argumentos de archivo, escanea el directorio actual automáticamente.

EOT;

$command = $argv[1] ?? null;
$rest    = array_slice($argv, 2);

if (!$command || $command === '-h' || $command === '--help') {
    echo $usage;
    exit(0);
}

match ($command) {
    'create' => cmd_create($rest),
    'open'   => cmd_open($rest),
    default  => (function () use ($command, $usage) {
        echo "  Comando desconocido: '{$command}'\n" . $usage;
        exit(1);
    })(),
};
